optuna/storages/_journal/file.py

```diff
@@ -164,28 +164,40 @@ def __init__(self, file_path: str, lock_obj: Optional[JournalFileBaseLock] = Non
     def read_logs(self, log_number_from: int) -> List[Dict[str, Any]]:
         logs = []
         with open(self._file_path, "rb") as f:
+            remaining_log_size = os.stat(self._file_path).st_size
             log_number_start = 0
             if log_number_from in self._log_number_offset:
                 f.seek(self._log_number_offset[log_number_from])
                 log_number_start = log_number_from
+                remaining_log_size -= self._log_number_offset[log_number_from]
 
             last_decode_error = None
             for log_number, line in enumerate(f, start=log_number_start):
                 if last_decode_error is not None:
                     raise last_decode_error
+                byte_len = len(line)
+                remaining_log_size -= byte_len
+                if remaining_log_size < 0:
+                    break
                 if log_number + 1 not in self._log_number_offset:
-                    byte_len = len(line)
                     self._log_number_offset[log_number + 1] = (
                         self._log_number_offset[log_number] + byte_len
                     )
                 if log_number < log_number_from:
                     continue
+                if line.decode("utf-8") in [
+                    "\n",
+                    "\r",
+                    "\r\n",
+                ]:  # to avoid json.loads(<line separator>)
```

Comment:
Can we just write this like below?

```suggestion
                if line in b"\r\n":  # to avoid json.loads(<line separator>)
```

Reply:
I found that this implementation and your suggestion did not work correctly and have fixed the implementation to handle the following corner case.

## Corner case

**Process 1 (writer)**
time step1-a: write '{"operation", "op1"}'
time step1-c: write '\n'
time step2: write '{"operation", "op2"}\n'
 
**Process 2 (reader)**
time step 1-b: (read) log_number = 0, line = '{"operation", "op1"}', offset[1] = len(line)
time step 3-a: (read) line = '\n'

**logfile** (after time step 2)
line 0: {"operation", "op1"}\n
line 1: {"operation", "op2"}\n


Based on my understanding, at time step 3-a, `log_number` should be 0 and `offset[1]` should be updated (`offset[1] += len('\n')`).
Then, reading should `continue` rather than `break` when finding a `line` only contains \<line separator\>), otherwise, an empty log will be returned nevertheless line 1 exists (the following time step 3-b).

**Process 2 (reader)**
time step 3-b: (read) log_number = 1, line = '{"operation", "op2"}\n', offset[1] = len('{"operation", "op1"}') + len('\n')

----

The following json.loads' behavior causes this corner case.
We need to avoid to evaluate `json.loads('\n')` but maintain the log_number and offset properly according to `'\n'`.

```python
>>> json.loads('{"hoge": "piyo"}')
{'hoge': 'piyo'}

>>> json.loads('{"hoge": "piyo"}\n')
{'hoge': 'piyo'}

>>> json.loads('\n')
json.decoder.JSONDecodeError: Expecting value: line 2 column 1 (char 1)
```

Reply:
Thank you for your feedback.

Your updated code's idea that ensures each line ends with linesep seems to work fine.
I think a line ends with '\r' should not be allowed because it will make a split like '{"op": "op1"}\r' and "\n".
And, platforms use '\r' (e.g. MacOS ver. < 9) as a linesep are quite rare today.